web/src/message_store.ts

```diff
@@ -1,82 +1,109 @@
+
+const message_reaction_type_schema = z.enum(["unicode_emoji", "realm_emoji", "zulip_extra_emoji"]);
+
+export type MessageReactionType = z.infer<typeof message_reaction_type_schema>;
+
+const display_recipient_user_schema = z.object({
+    email: z.string(),
+    full_name: z.string(),
+    id: z.number(),
+    is_mirror_dummy: z.boolean(),
+});
+
+export type DisplayRecipientUser = z.infer<typeof display_recipient_user_schema>;
+
+const display_recipient_schema = z.union([z.string(), z.array(display_recipient_user_schema)]);
+
+export type DisplayRecipient = z.infer<typeof display_recipient_schema>;
+
+const message_edit_history_entry_schema = z.object({
+    user_id: z.nullable(z.number()),
+    timestamp: z.number(),
+    prev_content: z.optional(z.string()),
+    prev_rendered_content: z.optional(z.string()),
+    prev_rendered_content_version: z.optional(z.number()),
+    prev_stream: z.optional(z.number()),
+    prev_topic: z.optional(z.string()),
+    stream: z.optional(z.number()),
+    topic: z.optional(z.string()),
+});
+
+export type MessageEditHistoryEntry = z.infer<typeof message_edit_history_entry_schema>;
+
+const message_reaction_schema = z.object({
+    emoji_name: z.string(),
+    emoji_code: z.string(),
+    reaction_type: message_reaction_type_schema,
+    user_id: z.number(),
+});
+
+export type MessageReaction = z.infer<typeof message_reaction_schema>;
+
+export const submessage_schema = z.object({
+    id: z.number(),
+    sender_id: z.number(),
+    message_id: z.number(),
+    content: z.string(),
+    msg_type: z.string(),
+});
+
+export const raw_message_schema = z.intersection(
+    z.intersection(
+        z.object({
+            avatar_url: z.nullable(z.string()),
+            client: z.string(),
+            content: z.string(),
+            content_type: z.literal("text/html"),
+            display_recipient: display_recipient_schema,
+            edit_history: z.optional(z.array(message_edit_history_entry_schema)),
+            id: z.number(),
+            is_me_message: z.boolean(),
+            last_edit_timestamp: z.optional(z.number()),
+            reactions: z.array(message_reaction_schema),
+            recipient_id: z.number(),
+            sender_email: z.string(),
+            sender_full_name: z.string(),
+            sender_id: z.number(),
+            sender_realm_str: z.string(),
+            submessages: z.array(submessage_schema),
+            timestamp: z.number(),
+            flags: z.array(z.string()),
+        }),
+        z.discriminatedUnion("type", [
+            z.object({
+                type: z.literal("private"),
+            }),
+            z.object({
+                type: z.literal("stream"),
+                stream_id: z.number(),
+                // Messages that come from the server use `subject`.
+                // Messages that come from `send_message` use `topic`.
+                subject: z.optional(z.string()),
```

Comment:
Messages of type `private` recieved from the server also contain `subject` field. Its an empty array `""`

Reply:
Yeah, it's an empty string `""` but do we need to parse that?

Reply:
Yeah. Probably not!.